File: labs/05-memory-system/node/main.ts
import { createHash } from "crypto";
import { OpenAIClient, ChatMessage } from "../../../shared/config/openai_client";

let chromaModule: any;
try {
  chromaModule = require("chromadb");
} catch {
  chromaModule = null;
}

class ShortTermMemory {
  messages: ChatMessage[] = [];

  constructor(private maxMessages = 10) {}

  add(role: string, content: string): void {
    this.messages.push({ role: role as ChatMessage["role"], content });
    if (this.messages.length > this.maxMessages) {
      const keep = this.messages[0]?.role === "system" ? 1 : 0;
      this.messages = this.messages.slice(0, keep).concat(this.messages.slice(keep + 1));
    }
  }

  get(): ChatMessage[] {
    return this.messages.slice();
  }
}

class LongTermMemory {
  collection: any = null;

  constructor(private client: OpenAIClient, collectionName = "agent_memory") {
    if (chromaModule?.ChromaClient) {
      const db = new chromaModule.ChromaClient({ path: "http://localhost:8000" });
      db.getOrCreateCollection({ name: collectionName }).then((col: any) => {
        this.collection = col;
      }).catch((err: any) => {
        console.warn("[Long-term memory] Could not connect to Chroma:", err.message);
      });
    }
  }

  async embed(text: string): Promise<number[]> {
    const resp = await this.client.chatCompletion({
      messages: [
        {
          role: "user",
          content: `Summarize in one sentence for retrieval: ${text}`,
        },
      ],
      max_tokens: 20,
    });
    const summary = resp.choices[0].message.content;
    const digest = createHash("sha256").update(summary).digest();
    const vec = new Array(64).fill(0);
    for (let i = 0; i < digest.length; i++) {
      vec[i % 64] += digest[i] / 255.0;
    }
    return vec;
  }

  async store(text: string, metadata?: Record<string, any>): Promise<void> {
    if (!this.collection) {
      console.log("[Long-term memory] Chroma not available, skipping store.");
      return;
    }
    const docId = createHash("sha256").update(text).digest("hex").slice(0, 16);
    await this.collection.add({
      ids: [docId],
      documents: [text],
      metadatas: [metadata || {}],
      embeddings: [await this.embed(text)],
    });
  }

  async retrieve(query: string, n = 3): Promise<string[]> {
    if (!this.collection) return [];
    const results = await this.collection.query({
      queryEmbeddings: [await this.embed(query)],
      nResults: n,
    });
    return results.documents?.[0] || [];
  }
}

class AgentWithMemory {
  shortTerm: ShortTermMemory;
  longTerm: LongTermMemory;

  constructor(private client: OpenAIClient) {
    this.shortTerm = new ShortTermMemory(12);
    this.longTerm = new LongTermMemory(client);
  }

  async chat(userInput: string): Promise<string> {
    const relevant = await this.longTerm.retrieve(userInput);
    if (relevant.length > 0) {
      const context = "Relevant memory:\n" + relevant.map((r) => `- ${r}`).join("\n");
      this.shortTerm.add("system", context);
    }

    this.shortTerm.add("user", userInput);
    const messages = this.shortTerm.get();
    const response = await this.client.chatCompletion({ messages, max_tokens: 200 });
    const answer = response.choices[0].message.content;
    this.shortTerm.add("assistant", answer);
    await this.longTerm.store(`User: ${userInput}\nAssistant: ${answer}`);
    return answer;
  }
}

async function main() {
  const client = new OpenAIClient();
  const agent = new AgentWithMemory(client);

  console.log("Agent:", await agent.chat("My name is Alice and I work on backend systems."));
  console.log("Agent:", await agent.chat("What do I work on?"));
  console.log("Agent:", await agent.chat("Suggest a logging strategy for my team."));
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: labs/05-memory-system/python/main.py
import hashlib
import json
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[3] / "shared" / "config"))

from openai_client import OpenAIClient, configure_logging


try:
    import chromadb
except ImportError:
    chromadb = None


class ShortTermMemory:
    def __init__(self, max_messages: int = 10):
        self.messages = []
        self.max_messages = max_messages

    def add(self, role: str, content: str) -> None:
        self.messages.append({"role": role, "content": content})
        if len(self.messages) > self.max_messages:
            # Keep system message if present, otherwise trim oldest
            keep = 1 if self.messages and self.messages[0]["role"] == "system" else 0
            self.messages = self.messages[:keep] + self.messages[keep + 1 :]

    def get(self) -> list:
        return list(self.messages)


class LongTermMemory:
    def __init__(self, client: OpenAIClient, collection_name: str = "agent_memory"):
        self.client = client
        self.collection = None
        if chromadb:
            db = chromadb.HttpClient(host="localhost", port=8000)
            self.collection = db.get_or_create_collection(collection_name)

    def embed(self, text: str) -> list[float]:
        # Use the chat model endpoint as a cheap proxy; replace with real embedding endpoint in production.
        resp = self.client.chat_completion(
            messages=[{"role": "user", "content": f"Summarize in one sentence for retrieval: {text}"}],
            max_tokens=20,
        )
        # In production use an embeddings endpoint. Here we hash to a fixed-size vector for demo compatibility.
        summary = resp["choices"][0]["message"]["content"]
        vec = [0.0] * 64
        for i, byte in enumerate(hashlib.sha256(summary.encode()).digest()):
            vec[i % 64] += byte / 255.0
        return vec

    def store(self, text: str, metadata: dict | None = None) -> None:
        if not self.collection:
            print("[Long-term memory] Chroma not available, skipping store.")
            return
        doc_id = hashlib.sha256(text.encode()).hexdigest()[:16]
        self.collection.add(
            ids=[doc_id],
            documents=[text],
            metadatas=[metadata or {}],
            embeddings=[self.embed(text)],
        )

    def retrieve(self, query: str, n: int = 3) -> list[str]:
        if not self.collection:
            return []
        results = self.collection.query(query_embeddings=[self.embed(query)], n_results=n)
        return results["documents"][0] if results["documents"] else []


class AgentWithMemory:
    def __init__(self, client: OpenAIClient):
        self.client = client
        self.short_term = ShortTermMemory(max_messages=12)
        self.long_term = LongTermMemory(client)

    def chat(self, user_input: str) -> str:
        relevant = self.long_term.retrieve(user_input)
        if relevant:
            context = "Relevant memory:\n" + "\n".join(f"- {r}" for r in relevant)
            self.short_term.add("system", context)

        self.short_term.add("user", user_input)
        messages = self.short_term.get()
        response = self.client.chat_completion(messages=messages, max_tokens=200)
        answer = response["choices"][0]["message"]["content"]
        self.short_term.add("assistant", answer)
        self.long_term.store(f"User: {user_input}\nAssistant: {answer}")
        return answer


def main():
    configure_logging()
    client = OpenAIClient()
    agent = AgentWithMemory(client)

    print("Agent:", agent.chat("My name is Alice and I work on backend systems."))
    print("Agent:", agent.chat("What do I work on?"))
    print("Agent:", agent.chat("Suggest a logging strategy for my team."))


if __name__ == "__main__":
    main()
